tests/chainermn_tests/functions_tests/test_point_to_point_communication.py

```diff
@@ -12,221 +11,251 @@
-            err = self.evaluation(z, self.x)
-            err.backward()
-
-        else:
-            y = chainermn.functions.recv(self.communicator, self.rank_recv)
-            err = chainermn.functions.send(
-                y, self.communicator, self.rank_send)
-            err.backward()
-
-    def test_tuple_communication1_cpu(self):
-        self.setup(False)
-        self.check_tuple_communication(1)
-
-    def test_tuple_communication2_cpu(self):
-        self.setup(False)
-        self.check_tuple_communication(2)
-
-    @chainer.testing.attr.gpu
-    def test_tuple_communication1_gpu(self):
-        self.setup(True)
-        self.check_tuple_communication(1)
-
-    @chainer.testing.attr.gpu
-    def test_tuple_communication2_gpu(self):
-        self.setup(True)
-        self.check_tuple_communication(2)
-
-
-class TestNonVariableInput(unittest.TestCase):
-
-    def setUp(self):
-        self.communicator = chainermn.create_communicator('naive')
-
-        if self.communicator.size < 2:
-            pytest.skip('This test is for multinode')
-
-        self.rank_send = (self.communicator.rank + 1) % self.communicator.size
-        self.rank_recv = (self.communicator.rank - 1) % self.communicator.size
-
-    def test_non_variable_send(self):
-        """Checks if backward will be called even if inputs are not Variable.
-
-        This test confirms whether deadlock occurs when numpy/cupy array is
-        given as an input of send.
-        In this case, the input will be converted to chainer Variable without
-        ``requires_grad``, thus ``backward`` will not be called without any
-        modification.
-        """
-        if self.communicator.rank == 0:
-            x = numpy.ones((1, 10)).astype(numpy.float32)
-            phi = chainermn.functions.send(
-                x, self.communicator, rank=self.rank_send)
-            x, = chainermn.functions.pseudo_connect(phi, x)
-            y = chainer.functions.sum(x)
-            t = numpy.array(0).astype(numpy.float32)
-            z = chainer.functions.mean_squared_error(y, t)
-            z.backward()
-
-        elif self.communicator.rank == self.communicator.size - 1:
-            x = chainermn.functions.recv(
-                self.communicator, rank=self.rank_recv)
-            y = chainer.functions.sum(x)
-            t = numpy.array(0).astype(numpy.float32)
-            z = chainer.functions.mean_squared_error(y, t)
-            z.backward()
-
-        else:
-            x = chainermn.functions.recv(
-                self.communicator, rank=self.rank_recv)
-            phi = chainermn.functions.send(
-                x, self.communicator, rank=self.rank_send)
-            phi.backward()
+
+params = [Param(p) for p in [
+    {
+        'dtype': numpy.float16,
+    }, {
+        'dtype': numpy.float32,
+    }]]
+
+
+def create_communicator(gpu, param):
+    if gpu:
+        communicator = chainermn.create_communicator('hierarchical')
```

Comment:
'hierarchical' is to be deprecated in next release. Use 'flat' instead for default communicator.